netrd/reconstruction/graphical_lasso.py

```diff
@@ -0,0 +1,149 @@
+
+        alpha (float, default=0.01): Coefficient of penalization, higher values
+        means more sparseness
+
+        max_iter (int, default=100): Maximum number of iterations.
+
+        convg_threshold (float, default=0.001): Stop the algorithm when the
+        duality gap is below a certain threshold.
+
+
+        Returns
+        -------
+        G (nx.Graph): A reconstructed graph with $N$ nodes.
+
+        """
+
+        cov, prec = graphical_lasso(TS, alpha, max_iter, convg_threshold)
+        G = nx.from_numpy_array(cov)
+        self.results['covariance'] = cov
+        self.results['precision'] = prec
+        self.results['graph'] = G
+
+        return G
+
+
+def graphical_lasso(TS, alpha=0.01, max_iter=100, convg_threshold=0.001):
+    """ This function computes the graphical lasso algorithm as outlined in [1].
+        
+    Params
+    ------
+    TS (np.ndarray): Array consisting of $L$ observations from $N$ sensors.
+
+    alpha (float, default=0.01): Coefficient of penalization, higher values
+    means more sparseness
+
+    convg_threshold (float, default=0.001): Stop the algorithm when the
+    duality gap is below a certain threshold.
+
+    Returns
+    -------
+    cov (np.ndarray): Estimator of the inverse covariance matrix with sparsity.
+    
+    """
+    
+    if alpha == 0:
+        return cov_estimator(TS)
+    n_features = TS.shape[1]
+
+    mle_estimate_ = cov_estimator(TS)
+    covariance_ = mle_estimate_.copy()
+    precision_ = np.linalg.pinv( mle_estimate_ )
+    indices = np.arange( n_features)
+    for i in xrange( max_iter):
+        for n in range( n_features ):
+            sub_estimate = covariance_[ indices != n ].T[ indices != n ]
+            row = mle_estimate_[ n, indices != n]
+            #solve the lasso problem
+            _, _, coefs_ = lars_path( sub_estimate, row, Xy = row, Gram = sub_estimate, 
+                                        alpha_min = alpha/(n_features-1.), copy_Gram = True,
+                                        method = "lars")
+            coefs_ = coefs_[:,-1] #just the last please.
+        #update the precision matrix.
+            precision_[n,n] = 1./( covariance_[n,n] 
+                                    - np.dot( covariance_[ indices != n, n ], coefs_  ))
+            precision_[indices != n, n] = - precision_[n, n] * coefs_
+            precision_[n, indices != n] = - precision_[n, n] * coefs_
+            temp_coefs = np.dot( sub_estimate, coefs_)
+            covariance_[ n, indices != n] = temp_coefs
+            covariance_[ indices!=n, n ] = temp_coefs
+        
+        #if test_convergence( old_estimate_, new_estimate_, mle_estimate_, convg_threshold):
+        if np.abs( _dual_gap( mle_estimate_, precision_, alpha ) )< convg_threshold:
+                break
+    else:
+        #this triggers if not break command occurs
+        print "The algorithm did not coverge. Try increasing the max number of iterations."
```

Comment:
It looks like this is implemented in python 2, not 3. We'll have to change it over.

Reply:
Sounds good!